### DISCUSSION TITLE
Help with isAdditionalFlowStep for Taint Flow Involving Pointers

### DISCUSSION THREAD

**Discussion by sei-yhindka** [github/codeql#16620] [Q&A]
Hello CodeQL Team,

We are trying to track taint flow of a pointer and its fields through a program. Specifically, we are trying to track the taint of the function parameter named `state` here: https://github.com/ggerganov/whisper.cpp/blob/master/whisper.cpp#L7159. Here is the query we are using:

```CodeQL
/**
 * @name Data Flow
 * @kind path-problem
 * @severity warning
 * @id cpp/data-flow
 */

import cpp
import semmle.code.cpp.dataflow.new.TaintTracking

module StructDataFlowConfiguration implements DataFlow::ConfigSig 
{
    predicate isSource(DataFlow::Node source) 
    {
        exists(Function fxn, Parameter p |

            // scope to function
            fxn.getName() = "whisper_exp_compute_token_level_timestamps_dtw"

            // scope to parameter within function
            and p.getFunction() = fxn
            and p.getName() = "state"
            and source.asParameter() = p

        )
    }

    predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) 
    {
        exists(PointerFieldAccess pfa |
            
            // check node2 is the FieldAccess
            node2.asExpr() = pfa

            // check node1 is an indirect parent
            and node1.asIndirectExpr() = pfa.getQualifier()
            
            // prevent false positives
            and pfa.getQualifier().getType().(UserType).hasName("whisper_state")

        )
    }
    
    predicate isSink(DataFlow::Node sink)
    {
        not isSource(sink) // Eliminate zero-step paths
    }
}

module StructDataFlow = TaintTracking::Global<StructDataFlowConfiguration>;
import StructDataFlow::PathGraph

from StructDataFlow::PathNode source, StructDataFlow::PathNode sink
// provenance is related to features from CodeQL for C# that have not been implemented for C++
where StructDataFlow::PathGraph::edges(source, sink, "provenance", "")
select 
    sink.getNode(), source, sink, "Found edge"
```

Currently, this query does track taint through all uses of the `state` parameter. However, it does not consider the fields of the `state` parameter as tainted, which is something we need for our use case. We have a feeling our issue is with the `isAdditionalFlowStep` predicate. How can we modify this so that fields of a pointer are considered tainted? Thanks in advance for your help!

**Comment by aibaars** [comment]
If I understand correctly you would like to add additional taint steps for field accesses that are performed on a tainted qualifier. 

I see you used an `or` in your definition of `isAdditionalTaintStep` that should probably be an `and`, otherwise you end up with very many false flow steps. I guess you added the additional restriction on the type to reduce the number of false flows.

To debug things like this I'd recommend using the "quick evaluation" feature from QL for VSCode. This feature allows you to select a predicate or expression, right click, to evaluate and inspect the results. This makes it easier to figure out where you are unexpectedly losing or gaining results.

**Reply by sei-yhindka** [reply]
Apologies, the OR was a typo; it was not present in the query I ran. I have edited the post to be accurate now.

The `isAdditionalFlowStep` predicate is for adding fields of structs as tainted variables. CodeQL does not regard fields of structs as tainted out-of-the-box. While the query works as intended for non-pointers to structs (i.e., by changing the `PointerFieldAccess` to `FieldAccess`, it does not work for pointers to structs.

**Reply by aibaars** [reply]
I suspect what is happening is that for a pointer field access the qualifier is the pointer and not the struct that the pointer refers to. I think your implementation of isAdditionalFlow would propagate the taint in cases where the pointer itself is tainted. It's a bit rare to have an untrusted user input flow directly into a pointer, but it possible, for example: `int index = ask_the_user(); int *p = start + index`. I think the case you are after is when there is pointer that refers to a struct containing tainted values and treat a pointer field reference into that struct as tainted.

Unfortunately, I am not too  familiar with how pointers are modelled in the QL library,  so I can't give you a solution to your problem.

**Reply by jketema** [reply]
If you want to taint the struct and not the pointer to the struct, then for the source you can use:
```ql
source.asParameter(1) = p
```
instead of
```ql
source.asParameter() = p
```
The documentation of `asParameter` explains the meaning of the `1` value: https://github.com/github/codeql/blob/61593aed7df93c254de2b4b2515e1ad2a21ef981/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L384-L397

**Reply by sei-yhindka** [reply]
Thanks so much to both of you, @aibaars and @jketema! We tried out your solution, @jketema, but it produced many false positives. We don't want to taint the struct itself, because then all instances of the struct are tainted, whereas we are only interested in the specific parameter (`state`) of the function. The query below is very close to achieving the results we want:

```CodeQL
exists(FieldAccess fa |
            
            // check node2 is the FieldAccess
            node2.asExpr() = fa

            // check node1 is an indirect parent
            and node1.asIndirectExpr() = fa.getQualifier()
            
            // prevent false positives
            and fa.getQualifier().getType().(PointerType).getBaseType().hasName("whisper_state")

        )
```

When we run the quick eval on the `isSource` predicate, CodeQL correctly identifies the `state` parameter as the only source. In addition, when we run the quick eval on the `isAdditionalFlowStep` predicate, CodeQL correctly identifies the field accesses of the the `state` parameter as tainted (see the screenshot below). However, when we run the whole query, we do not get the results we expect: the field accesses of the `state` parameter are not part of the results. We have a feeling there is a disconnect between the source and the nodes selected by the `isAdditionalFlowStep` predicate. Do you have any ideas on how to proceed?

<img width="1507" alt="Screenshot 2024-05-30 at 3 49 03 PM" src="https://github.com/github/codeql/assets/144383669/b95de384-530d-4e1f-924a-9aac4a669dde">

**Reply by jketema** [reply]
@sei-yhindka I'm not sure I understand what you're actually after. On the one hand it seems that you only want the `state` pointer to be tainted. On the other hand it seems you want the whole struct to be tainted, because it seems you're trying to add an addition step that taints all the field accesses. Those seem to be conflicting requirements.

**Reply by sei-yhindka** [reply]
@jketema thank you for your response and we apologize for any confusion. We are trying to follow the taint of the `state` parameter, including its fields. So, as you can see on the right side of the screenshot above, CodeQL identifies flow from the pointer to its fields (which is what we want). However, those results are from a quick evaluation of the `isAdditionalFlowStep` predicate. When we run the whole query (including `isSource` and `isSink` predicates), we do not see these same flows, leading us to believe that there is a disconnect between the results of the `isSource` predicate and the results of the `isAdditionalFlowStep` predicate. This may be easier to resolve over a video conference call, so please let us know if that is an option.

**Reply by jketema** [reply]
Thanks for the clarification. In that case I do not understand why the solution I proposed above does not work for you.
